packages/core/src/Machine.ts

```diff
@@ -50,23 +58,64 @@ export function Machine<
 export function createMachine<
   TContext,
   TEvent extends EventObject = AnyEventObject,
-  TTypestate extends Typestate<TContext> = { value: any; context: TContext }
+  TTypestate extends Typestate<TContext> = { value: any; context: TContext },
+  TTypesMeta extends TypegenConstraint = TypegenDisabled,
+  TResolvedTypesMeta = ResolveTypegenMeta<TTypesMeta, TEvent, BaseActionObject>
 >(
   config: TContext extends Model<any, any, any, any>
     ? 'Model type no longer supported as generic type. Please use `model.createMachine(...)` instead.'
-    : MachineConfig<TContext, any, TEvent>,
-  options?: Partial<MachineOptions<TContext, TEvent>>
-): StateMachine<TContext, any, TEvent, TTypestate>;
+    : MachineConfig<TContext, any, TEvent, BaseActionObject, TTypesMeta>,
+  options?: MaybeTypegenMachineOptions<
+    TContext,
+    TEvent,
+    BaseActionObject,
+    TResolvedTypesMeta
+  >
+): StateMachine<
+  TContext,
+  any,
+  TEvent,
+  any,
+  BaseActionObject,
+  TResolvedTypesMeta
+>;
+
 export function createMachine<
   TContext,
   TEvent extends EventObject = AnyEventObject,
-  TTypestate extends Typestate<TContext> = { value: any; context: TContext }
+  TTypestate extends Typestate<TContext> = { value: any; context: TContext },
+  TTypesMeta extends TypegenConstraint = TypegenDisabled
 >(
-  config: MachineConfig<TContext, any, TEvent>,
-  options?: Partial<MachineOptions<TContext, TEvent>>
-): StateMachine<TContext, any, TEvent, TTypestate> {
+  config: MachineConfig<
+    TContext,
+    any,
+    TEvent,
+    ActionObject<TContext, TEvent>,
+    TTypesMeta
+  >,
+  options?: MaybeTypegenMachineOptions<
+    TContext,
+    TEvent,
+    BaseActionObject,
+    TTypesMeta
+  >
+): StateMachine<
+  TContext,
+  any,
+  TEvent,
+  TTypestate,
+  BaseActionObject,
+  TTypesMeta
+> {
   return new StateNode<TContext, any, TEvent, TTypestate>(
     config,
-    options
-  ) as StateMachine<TContext, any, TEvent, TTypestate>;
+    options as any
```

Comment:
What needs to be fixed here?

Reply:
What do you mean?

Reply:
The options type is basically too dynamic now - it can either be specified loosely or tightly, depending on whether we have generated types. Typescript isn't smart enough, so we patch it with any.

Reply:
Actually, after some changes, this cast is no longer needed. 🤷‍♂️ I've removed it locally